src/man/man3/catcmds.3
.TH catcmds 3U "23 January 1991" "Unidraw" "InterViews Reference Manual"
.SH NAME
NewCompCmd, PrintCmd, QuitCmd, RevertCmd, SaveCompCmd, SaveCompAsCmd,
ViewCompCmd \- catalog-oriented commands
.SH SYNOPSIS
.B #include <Unidraw/Commands/catcmds.h>
.SH DESCRIPTION
The Unidraw library predefines a set of non-interpretive commands that
let a user specify components to open, view, and save, plus commands
for generating a PostScript external representation and for quitting
the application.  NewCompCmd re-initializes its editor's component to
let the user start editing afresh.  If the existing component is
modified, NewCompCmd prompts the user to save it.  PrintCmd generates
a PostScriptView of the editor's component and lets the user send it
to a file or to a printer.  QuitCmd quits the application, prompting
the user to save modified components.  RevertCmd lets the user re-open
the current component, discarding changes since it was last saved.
SaveCompAsCmd lets the user save a component under a particular name,
while SaveCompCmd saves a component named previously.  Finally,
ViewCompCmd lets the user specify another component to view in its
editor.
.SH NEWCOMPCMD PUBLIC OPERATIONS
.TP
.B "NewCompCmd(ControlInfo*, Component* prototype = nil)"
.ns
.TP
.B "NewCompCmd(Editor* = nil, Component* = nil)"
Create a new NewCompCmd instance, optionally supplying a prototype
component.  When executed, NewCompCmd will make its editor edit a copy
of the prototype.  The editor's original component will be deleted if
no other editor refers to it.
.TP
.B "virtual boolean Reversible()"
NewCompCmd is not reversible.
.TP
.B "Component* GetPrototype()"
Return the prototype component specified in the constructor.
.SH PRINTCMD PUBLIC OPERATIONS
.TP
.B "PrintCmd(ControlInfo*, PrintDialog* = nil)"
.ns
.TP
.B "PrintCmd(Editor* = nil, PrintDialog* = nil)"
Create a new PrintCmd instance, optionally supplying the PrintDialog
to display when executed.
.TP
.B "virtual boolean Reversible()"
PrintCmd is not reversible.
.SH PRINTCMD PROTECTED OPERATIONS
.TP
.B "void print(const char* print_cmd, const char* file)"
Concatenates \fIprint_cmd\fP and \fIfile\fP and calls system(3) on the
result.
.SH QUITCMD PUBLIC OPERATIONS
.TP
.B "QuitCmd(ControlInfo*)"
.ns
.TP
.B "QuitCmd(Editor* = nil)"
Create a new QuitCmd instance.
.TP
.B "virtual boolean Reversible()"
QuitCmd is not reversible.
.SH REVERTCMD PUBLIC OPERATIONS
.TP
.B "RevertCmd(ControlInfo*)"
.ns
.TP
.B "RevertCmd(Editor* = nil)"
Create a new RevertCmd instance.
.TP
.B "virtual boolean Reversible()"
RevertCmd is not reversible.
.SH SAVECOMPCMD PUBLIC OPERATIONS
.TP
.B "SaveCompCmd(ControlInfo*)"
.ns
.TP
.B "SaveCompCmd(Editor* = nil)"
Create a new SaveCompCmd instance.
.TP
.B "virtual boolean Reversible()"
SaveCompCmd is not reversible.
.SH SAVECOMPASCMD PUBLIC OPERATIONS
.TP
.B "SaveCompAsCmd(ControlInfo*, FileChooser* = nil)"
.ns
.TP
.B "SaveCompAsCmd(Editor* = nil, FileChooser* = nil)"
Create a new SaveCompAsCmd instance, optionally supplying the
FileChooser that will prompt the user to name the component.
.TP
.B "virtual boolean Reversible()"
SaveCompAsCmd is not reversible.
.SH VIEWCOMPCMD PUBLIC OPERATIONS
.TP
.B "ViewCompCmd(ControlInfo*, FileChooser* = nil)"
.ns
.TP
.B "ViewCompCmd(Editor* = nil, FileChooser* = nil)"
Create a new ViewCompCmd instance, optionally supplying the
FileChooser that will prompt the user to specify the component to
view.  After the new component is opened successfully, the editor's
original component will be deleted if no other editor refers to it.
.TP
.B "virtual boolean Reversible()"
ViewCompCmd is not reversible.
.SH SEE ALSO
BasicDialog(3U), Command(3U), Component(3U), Editor(3U),
PostScriptView(3U), lpr(1).


CHANGES
July 9th, 2008  ivtools-1.2.6

- Debian contributed patch for iostreams backward compatibility

- bug fix for dispatcher.c contributed by Damon Permezel

- add mute command to comterp

- balance of changes to support ipl-1.0.2

March 4th, 2008  ivtools-1.2.5

- overall evolution to support use by ipl-1.0.1 (see
  http://sf.net/projects/ipl for further details)

July 20th, 2006  ivtools-1.2.4

- changes for compiling with gcc-4.0

- changes to socket-driven (ACE) use of comterp to support multi-line commands

- changes to allow re-use of scanner/parser with multiple operator tables

October 7th, 2005  ivtools-1.2.3

- add support for an external libtiff with a --with-tiff (and
-with-tiff-lib) configure arguments.

- changes necessary to compile with gcc-4.0

- changes necessary to compile with gcc-4.0 on top of ACE-5.4.7

- changes necessary to compile on suse-9.3 linux (amd64 platform).
Contributed by Paul Vixie.

- fixed bug in "addtool" command/mechanism in graphdraw and drawserv

- added support for color expressions to graphdraw (:colexpr), to
allow attribute-sensitive computation of a node's color.

January 25th, 2005  ivtools-1.2.2

If you submitted a patch in the past half year to ivtools that you
don't see reflected here, please resubmit.  I lost that e-mail.

- fix for problem writing to /tmp

- compilation fix on NetBSD for comterp.c

- compilation patch submitted by Michal Palczew.

July 13th, 2004  ivtools-1.2.1

Two crashes discovered and fixed by Michal Palczewski (mpalczew at
users.sourceforge.net).  One was a crash on exit after a delete with
idraw/drawtool, the other was a crash upon delete with graphdraw.
They turned out to be long-standing bugs in the order of object
deletion, problems that hadn't appeared with other versions of gcc.
RedHat must have adopted a more aggressive recycling of freed space in
a recent release.  Thanks Michal.


June 3rd, 2004  ivtools-1.2

ivtools-1.2 sets a new direction for the package, toward peer-to-peer
drawing chat.  ivtools' drawserv, the editor that incorporates every ivtool
mechanism to date (multi-frame, graph-editing, integrated scripting
language) has been extended to support a unique kind of
small-granularity shared whiteboard.  

This work was based on the pre-existing networked scripting capability
of ivtools' comdraw.  Then the pasting of any graphic was intercepted,
and the graphic serialized into a command that could be communicated
to other drawing editors to keep them in sync.

Then a connection management system was established, that supported
two-way peer-to-peer links between drawing editors.  In addition the
return path on each socket was used for detecting timeout, which
causes the connection to be dropped.

Finally, a distributed mutually-exclusive graphic selection mechanism
was created, where every drawing editor is aware of who currently has
a graphic selected or who had it selected last if it is not currently
selected.  When one drawing editor tries to select a graphic, it first
requests that permission from the last selector (after determining it
is unselected).

Now you can link to other drawing editors, draw graphics and see them
appear remotely (although only in black and white), and exercise the
distributed selection mechanism.  Four pieces remain to be developed:
1) support for full-up graphic state transmittal (it is sent, but not
interpreted on the remote end);, 2) support for transmission of the
complete set of drawing editor commands; 3) extending the distributed
mutual-exclusion mechanism to select pseudo-graphics to acquire global
permission for connection, pasting, moving front-to-back, and
grouping; and 4) a two-tier mechanism for invoking select(), that will
allow the bypass of selection requests while engaged in lengthy direct
manipulations.

If all this interests you, and you'd like to help, build this version
of ivtools with ACE support, subscribe to the ivtools-dev mailing, and
help me do some testing, programming, thinking, whatever.  Or visit
http://www.ivtools.org to make a donation to support this work.  Thanks.

Scott Johnston




src/Unidraw/catcmds.c
/*
 * Copyright (c) 1990, 1991 Stanford University
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Stanford not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  Stanford makes no representations about
 * the suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * STANFORD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.
 * IN NO EVENT SHALL STANFORD BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Implementation of cataloging commands.
 */

#include <Unidraw/catalog.h>
#include <Unidraw/classes.h>
#include <Unidraw/dialogs.h>
#include <Unidraw/editor.h>
#include <Unidraw/iterator.h>
#include <Unidraw/statevars.h>
#include <Unidraw/unidraw.h>
#include <Unidraw/Commands/catcmds.h>
#include <Unidraw/Components/grcomp.h>
#include <Unidraw/Components/grview.h>
#include <Unidraw/Components/psview.h>

#include <IV-look/dialogs.h>
#include <InterViews/session.h>
#include <InterViews/style.h>
#include <InterViews/window.h>
#include <OS/string.h>

#include <osfcn.h>
#include <stdio.h>
#include <stdlib.h>
#include <stream.h>
#include <string.h>

/*****************************************************************************/

static void UpdateCompNameVars () {
    Iterator i;

    for (unidraw->First(i); !unidraw->Done(i); unidraw->Next(i)) {
        Editor* ed = unidraw->GetEditor(i);
        CompNameVar* compNameVar = (CompNameVar*) ed->GetState("CompNameVar");
        if (compNameVar != nil) compNameVar->UpdateName();
    }
}

static boolean Writable (Component* comp) {
    Catalog* catalog = unidraw->GetCatalog();
    const char* name = catalog->GetName(comp);
    return name == nil || (catalog->Exists(name) && catalog->Writable(name));
}

static boolean OnlyOneEditorOf (Component* c) {
    Component* comp = c->GetRoot();
    Iterator i;
    int count = 0;

    for (unidraw->First(i); !unidraw->Done(i) && count < 2; unidraw->Next(i)) {
        Component* test_comp = unidraw->GetEditor(i)->GetComponent();

        if (test_comp != nil && test_comp->GetRoot() == comp) {
            ++count;
        }
    }
    return count == 1;
}

static boolean ReadyToClose (Editor* ed) {
    ModifStatusVar* mv = (ModifStatusVar*) ed->GetState("ModifStatusVar");

    if (mv != nil && Writable(mv->GetComponent()) && mv->GetModifStatus()) {
        ConfirmDialog dialog("Save changes?");

        ed->InsertDialog(&dialog);
        char resp = dialog.Confirm();
        ed->RemoveDialog(&dialog);

        if (resp == '\007') {
            return false;                       // confirm dialog aborted

        } else if (resp == 'y') {
            SaveCompCmd saveComp(ed);
            saveComp.Execute();

            if (mv->GetModifStatus()) {
                return false;                   // save dialog was aborted
            }
        }
    }
    return true;
}

/*****************************************************************************/

ClassId NewCompCmd::GetClassId () { return NEWCOMP_CMD; }

boolean NewCompCmd::IsA (ClassId id) {
    return NEWCOMP_CMD == id || Command::IsA(id);
}

NewCompCmd::NewCompCmd (ControlInfo* c, Component* p) : Command(c) { 
    prototype_ = p; 
}

NewCompCmd::NewCompCmd (Editor* ed, Component* p) : Command(ed) { 
    prototype_ = p; 
}

NewCompCmd::~NewCompCmd () { delete prototype_; }

Command* NewCompCmd::Copy () {
    Command* copy = new NewCompCmd(CopyControlInfo(), prototype_->Copy());
    InitCopy(copy);
    return copy;
}

void NewCompCmd::Execute () {
    Editor* ed = GetEditor();
    Component* orig = ed->GetComponent();
    Component* comp = prototype_->Copy();
    CompNameVar* compNameVar = (CompNameVar*) ed->GetState("CompNameVar");
    ModifStatusVar* modifVar = (ModifStatusVar*)ed->GetState("ModifStatusVar");

    if (OnlyOneEditorOf(orig) && !ReadyToClose(ed)) {
        return;
    }

    if (compNameVar != nil) compNameVar->SetComponent(comp);
    if (modifVar != nil) modifVar->SetComponent(comp);

    ed->SetComponent(comp);
    ed->Update();

    if (orig != nil && unidraw->FindAny(orig) == nil) {
        Component* root = orig->GetRoot();
        delete root;
    }
}

boolean NewCompCmd::Reversible () { return false; }

void NewCompCmd::Read (istream& in) {
    Command::Read(in);
    prototype_ = unidraw->GetCatalog()->ReadComponent(in);
}

void NewCompCmd::Write (ostream& out) {
    Command::Write(out); 
    unidraw->GetCatalog()->WriteComponent(prototype_, out);
}

/*****************************************************************************/

ClassId RevertCmd::GetClassId () { return REVERT_CMD; }

boolean RevertCmd::IsA (ClassId id) {
    return REVERT_CMD == id || Command::IsA(id);
}

RevertCmd::RevertCmd (ControlInfo* c) : Command(c) { }
RevertCmd::RevertCmd (Editor* ed) : Command(ed) { }

Command* RevertCmd::Copy () {
    Command* copy = new RevertCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void RevertCmd::Execute () {
    Editor* ed = GetEditor();
    Component* comp = ed->GetComponent();
    Catalog* catalog = unidraw->GetCatalog();
    const char* name = catalog->GetName(comp);
    ModifStatusVar* mv = (ModifStatusVar*) ed->GetState("ModifStatusVar");

    if (name != nil && (mv == nil || mv->GetModifStatus())) {
        char buf[CHARBUFSIZE];
        strcpy(buf, name);

        ConfirmDialog dialog("Really revert to last version saved?");
        ed->InsertDialog(&dialog);
        char confirmation = dialog.Confirm();
        ed->RemoveDialog(&dialog);

        if (confirmation == 'y') {
            Component* orig = comp;
            catalog->Forget(orig);

            if (unidraw->GetCatalog()->Retrieve(buf, comp)) {
                ed->SetComponent(comp);
                unidraw->CloseDependents(orig);
                unidraw->Update();

                CompNameVar* cv = (CompNameVar*) ed->GetState("CompNameVar");

                if (cv != nil) cv->SetComponent(comp);
                if (mv != nil) mv->SetComponent(comp);

                Component* root = orig->GetRoot();
                delete root;

            } else {
                ConfirmDialog dialog(
                    "Couldn't revert! (File nonexistent?)", "Save changes?"
                );
                ed->InsertDialog(&dialog);
                char confirmation = dialog.Confirm();
                ed->RemoveDialog(&dialog);

                UpdateCompNameVars();
                if (mv != nil) mv->Notify();

                if (confirmation == 'y') {
                    SaveCompAsCmd saveCompAs(ed);
                    saveCompAs.Execute();
                }
            }
        }
    }
}

boolean RevertCmd::Reversible () { return false; }

/*****************************************************************************/

ClassId ViewCompCmd::GetClassId () { return VIEWCOMP_CMD; }

boolean ViewCompCmd::IsA (ClassId id) {
    return VIEWCOMP_CMD == id || Command::IsA(id);
}

ViewCompCmd::ViewCompCmd (ControlInfo* c, FileChooser* fc) : Command(c) {
    chooser_ = fc;
    Resource::ref(chooser_);
}

ViewCompCmd::ViewCompCmd (Editor* ed, FileChooser* fc) : Command(ed) {
    chooser_ = fc;
    Resource::ref(chooser_);
}

ViewCompCmd::~ViewCompCmd () {
    Resource::unref(chooser_);
}

Command* ViewCompCmd::Copy () {
    Command* copy = new ViewCompCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void ViewCompCmd::Execute () {
    Editor* ed = GetEditor();

    if (OnlyOneEditorOf(ed->GetComponent()) && !ReadyToClose(ed)) {
        return;
    }

    Style* style;
    boolean reset_caption = false;
    if (chooser_ == nil) {
	style = new Style(Session::instance()->style());
	chooser_ = DialogKit::instance()->file_chooser(".", style);
	Resource::ref(chooser_);
	char buf[CHARBUFSIZE];
	const char* domain = unidraw->GetCatalog()->GetAttribute("domain");
	domain = (domain == nil) ? "component" : domain;
	sprintf(buf, "Select a %s to open:", domain);
	style->attribute("caption", "");
	style->attribute("subcaption", buf);
    } else {
	style = chooser_->style();
    }
    while (chooser_->post_for(ed->GetWindow())) {
        const String* s = chooser_->selected();
	NullTerminatedString ns(*s);
	const char* name = ns.string();
        Catalog* catalog = unidraw->GetCatalog();
        GraphicComp* comp;

        if (catalog->Retrieve(name, (Component*&) comp)) {
	    ModifStatusVar* modif = (ModifStatusVar*) ed->GetState(
		"ModifStatusVar"
	    );
            Component* orig = ed->GetComponent();
            ed->SetComponent(comp);
            unidraw->Update();

            StateVar* sv = ed->GetState("CompNameVar");
            CompNameVar* cnv = (CompNameVar*) sv;

            if (cnv != nil) cnv->SetComponent(comp);
            if (modif != nil) modif->SetComponent(comp);

            if (orig != nil && unidraw->FindAny(orig) == nil) {
                Component* root = orig->GetRoot();
                delete root;
            }
	    break;
        } else {
	    chooser_->bodyclear();
	    style->attribute("caption", "Open failed!");
	    reset_caption = true;
        }
    }
    if (reset_caption) {
      chooser_->bodyclear();
	style->attribute("caption", "");
    }
}

boolean ViewCompCmd::Reversible () { return false; }

/*****************************************************************************/

ClassId SaveCompCmd::GetClassId () { return SAVECOMP_CMD; }

boolean SaveCompCmd::IsA (ClassId id) {
    return SAVECOMP_CMD == id || Command::IsA(id);
}

SaveCompCmd::SaveCompCmd (ControlInfo* c) : Command(c) { }
SaveCompCmd::SaveCompCmd (Editor* ed) : Command(ed) { }

Command* SaveCompCmd::Copy () {
    Command* copy = new SaveCompCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void SaveCompCmd::Execute () {
    Editor* ed = GetEditor();
    ModifStatusVar* modifVar = (ModifStatusVar*)ed->GetState("ModifStatusVar");
    CompNameVar* compNameVar = (CompNameVar*) ed->GetState("CompNameVar");
    const char* name = (compNameVar == nil) ? nil : compNameVar->GetName();

    if (name == nil) {
        SaveCompAsCmd saveCompAs(ed);
        saveCompAs.Execute();

    } else if (modifVar == nil || modifVar->GetModifStatus()) {
        Catalog* catalog = unidraw->GetCatalog();
        Component* comp;

        if (catalog->Retrieve(name, comp) && catalog->Save(comp, name)) {
            if (modifVar != nil) modifVar->SetModifStatus(false);
            unidraw->ClearHistory(comp);

        } else {
            char title[CHARBUFSIZE];
            const char* reason = !Writable(comp) ? "(File not writable.)" : "";
            sprintf(title, "Couldn't save! %s", reason);

            char subtitle[CHARBUFSIZE];
            const char* domain = unidraw->GetCatalog()->GetAttribute("domain");
            domain = (domain == nil) ? "component" : domain;
            sprintf(subtitle, "Save this %s as:", domain);

	    Style* s = new Style(Session::instance()->style());
	    s->attribute("caption", title);
	    s->attribute("subcaption", subtitle);
	    s->attribute("open", "Save");
	    /* BUG: style s is never used!!!! */
            SaveCompAsCmd saveCompAs(ed);
            saveCompAs.Execute();
        }
    }
}

boolean SaveCompCmd::Reversible () { return false; }

/*****************************************************************************/

ClassId SaveCompAsCmd::GetClassId () { return SAVECOMPAS_CMD; }

boolean SaveCompAsCmd::IsA (ClassId id) {
    return SAVECOMPAS_CMD == id || Command::IsA(id);
}

SaveCompAsCmd::SaveCompAsCmd (ControlInfo* c, FileChooser* fc) : Command(c) {
    chooser_ = fc;
    Resource::ref(chooser_);
}

SaveCompAsCmd::SaveCompAsCmd (Editor* ed, FileChooser* fc) : Command(ed) {
    chooser_ = fc;
    Resource::ref(chooser_);
}

SaveCompAsCmd::~SaveCompAsCmd () {
    Resource::unref(chooser_);
}

Command* SaveCompAsCmd::Copy () {
    Command* copy = new SaveCompAsCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void SaveCompAsCmd::Execute () {
    Editor* ed = GetEditor();

    char buf[CHARBUFSIZE];
    const char* domain = unidraw->GetCatalog()->GetAttribute("domain");
    domain = (domain == nil) ? "component" : domain;
    sprintf(buf, "Save this %s as:", domain);

    boolean reset_caption = false;
    Style* style = new Style(Session::instance()->style());
    style->attribute("subcaption", buf);
    style->attribute("open", "Save");
    if (chooser_ == nil) {
	chooser_ = DialogKit::instance()->file_chooser(".", style);
	Resource::ref(chooser_);
    }
    while (chooser_->post_for(ed->GetWindow())) {
	const String* str = chooser_->selected();
	NullTerminatedString ns(*str);
        const char* name = ns.string();
        Catalog* catalog = unidraw->GetCatalog();
        boolean ok = true;

        if (catalog->Exists(name) && catalog->Writable(name)) {
            char buf[CHARBUFSIZE];
            sprintf(buf, "\"%s\" already exists.", name);
            ConfirmDialog dialog(buf, "Overwrite?");
            ed->InsertDialog(&dialog);
            char confirmation = dialog.Confirm();
            ed->RemoveDialog(&dialog);

            if (confirmation == 'n') {
                ok = false;
            } else if (confirmation != 'y') {
		break;
            }
        }
        if (ok) {
            CompNameVar* cnv = (CompNameVar*) ed->GetState("CompNameVar");
            const char* oldname = (cnv == nil) ? nil : cnv->GetName();
            Component* comp = ed->GetComponent();

            if (catalog->Exists(name) && !catalog->Writable(name)) {
		style->attribute(
		    "caption", "Couldn't save! (File not writable.)"
		);
            } else {
                if (oldname == nil) {
                    comp = comp->GetRoot();
                } else {
                    catalog->Retrieve(oldname, comp);
                    catalog->Forget(comp);
                }

                StateVar* sv = ed->GetState("ModifStatusVar");
                ModifStatusVar* mv = (ModifStatusVar*) sv;

                if (catalog->Save(comp, name)) {
                    if (mv != nil) mv->SetModifStatus(false);
                    unidraw->ClearHistory(comp);
                    UpdateCompNameVars();
		    break;
                } else {
                    if (mv != nil) mv->Notify();
                    UpdateCompNameVars();
		    style->attribute("caption", "Couldn't save!");
		    reset_caption = true;
                }
            } 
        }
    }
    if (reset_caption) {
	style->attribute("caption", "");
    }
}

boolean SaveCompAsCmd::Reversible () { return false; }

/*****************************************************************************/

ClassId PrintCmd::GetClassId () { return PRINT_CMD; }

boolean PrintCmd::IsA (ClassId id) {
    return PRINT_CMD == id || Command::IsA(id);
}

PrintCmd::PrintCmd (ControlInfo* c, PrintDialog* pd) : Command(c) {
    _dialog = pd;
}

PrintCmd::PrintCmd (Editor* ed, PrintDialog* pd) : Command(ed) {
    _dialog = pd;
}

PrintCmd::~PrintCmd () { delete _dialog; }

Command* PrintCmd::Copy () {
    Command* copy = new PrintCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void PrintCmd::Execute () {
    GraphicComp* comps = GetGraphicComp();
    boolean ok;

    if (_dialog == nil) {
        _dialog = new PrintDialog;
    }

    do {
        _editor->InsertDialog(_dialog);
        boolean accepted = _dialog->Accept();
        _editor->RemoveDialog(_dialog);

        if (!accepted) {
            break;
        }

        filebuf fbuf;
        char* tmpfilename;

        if (_dialog->ToPrinter()) {
            tmpfilename = tmpnam(nil);
            ok = fbuf.open(tmpfilename, output) != 0;
        } else {
            ok = fbuf.open((char*) _dialog->Choice(), output) != 0;
        }

        if (ok) {
            ostream out(&fbuf);
            ExternView* ev = (ExternView*) comps->Create(POSTSCRIPT_VIEW);
            comps->Attach(ev);
            ev->Update();
            ok = ev->Emit(out);
	    out.flush();
            delete ev;

            if (_dialog->ToPrinter()) {
                ok = print(_dialog->Choice(), tmpfilename) == 0;
            }
        }
        if (!ok) {
            _dialog->SetTitle("Couldn't print!");
        }
    } while (!ok);

    _dialog->SetTitle("");
}

int PrintCmd::print (const char* print_cmd, const char* file) {
    char cmd[CHARBUFSIZE];
    sprintf(cmd, "%s %s", print_cmd, file);
    return system(cmd);
}

boolean PrintCmd::Reversible () { return false; }

/*****************************************************************************/

ClassId QuitCmd::GetClassId () { return QUIT_CMD; }
boolean QuitCmd::IsA (ClassId id) { return QUIT_CMD == id || Command::IsA(id);}
QuitCmd::QuitCmd (ControlInfo* c) : Command(c) { }
QuitCmd::QuitCmd (Editor* ed) : Command(ed) { }

Command* QuitCmd::Copy () {
    Command* copy = new QuitCmd(CopyControlInfo());
    InitCopy(copy);
    return copy;
}

void QuitCmd::Execute () {
    Editor* ed = GetEditor();
    
    if (ReadyToClose(ed)) {
        Component* comp = ed->GetComponent();

        if (comp == nil) {
            unidraw->Close(ed);
        } else {
            unidraw->CloseDependents(comp->GetRoot());
        }
        Iterator i;

        for (;;) {
            unidraw->First(i);

            if (unidraw->Done(i)) {
                break;
            }

            ed = unidraw->GetEditor(i);

            if (ReadyToClose(ed)) {
                comp = ed->GetComponent();

                if (comp == nil) {
                    unidraw->Close(ed);
                } else {
                    unidraw->CloseDependents(comp->GetRoot());
                }
            } else {
                return;
            }
        }
        unidraw->Quit();
    }
}

boolean QuitCmd::Reversible () { return false; }
